src/Libraries/Microsoft.Extensions.AI/Utilities/AIJsonUtilities.Sse.cs

```diff
@@ -0,0 +1,105 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Text;
+using System.Text.Json;
+using System.Text.Json.Serialization.Metadata;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Shared.Diagnostics;
+using Microsoft.Shared.Pools;
+
+#pragma warning disable SA1114 // Parameter list should follow declaration
+
+namespace Microsoft.Extensions.AI;
+
+public static partial class AIJsonUtilities
+{
+    private static readonly byte[] _sseEventFieldPrefix = "event: "u8.ToArray();
+    private static readonly byte[] _sseDataFieldPrefix = "data: "u8.ToArray();
+    private static readonly byte[] _sseIdFieldPrefix = "id: "u8.ToArray();
+    private static readonly byte[] _sseLineBreak = Encoding.UTF8.GetBytes(Environment.NewLine);
+
+    /// <summary>
+    /// Serializes the specified server-sent events to the provided stream as JSON data.
+    /// </summary>
+    /// <typeparam name="T">Specifies the type of data payload in the event.</typeparam>
+    /// <param name="stream">The UTF-8 stream to write the server-sent events to.</param>
+    /// <param name="sseEvents">The events to serialize to the stream.</param>
+    /// <param name="options">The options configuring serialization.</param>
+    /// <param name="cancellationToken">The token taht can be used to cancel the write operation.</param>
+    /// <returns>A task representing the asynchronous write operation.</returns>
+    public static async ValueTask SerializeAsSseAsync<T>(
```

Comment:
If we want this as a public API, I'd rather it be in System.Net.ServerSentEvents.  For now it can be a non-public implementation detail from anything that needs to use it. I don't think we should be exposing this publicly from M.E.AI.

Reply:
Should I file a proposal in runtime following this shape?

Reply:
> Do you have any particular use case in mind? Like a specific streamer for IAsyncEnumerable<StreamingChatCompletionUpdate> or something else?

The use case I've been talking about all along, being able to write out the M.E.AI object model as an OpenAI-compatible response, both non-streaming and streaming varieties (this case being the latter).

> Should I file a proposal in runtime following this shape?

Sure. But in runtime ideally I'd want it to be something ASP.NET would rely on, so it'd be good to understand what its needs would be.